### PR TITLE
WT-2349  Initial addition of readonly configuration option.

### PR DESCRIPTION
**Pull request by sueloverso**
This is not yet ready for review.  It needs testing but config parsing (fairly complex) and incompatibility testing is in.

### CODE REVIEW

**Comment by sueloverso** [comment]
autobuild: retest this please

**Comment by sueloverso** [comment]
@michaelcahill or @agorrod This is ready for review now.  I have not yet added a readonly part to wtperf.

**Comment by michaelcahill** [comment]
@sueloverso, for your consideration, I pushed adb0461 to clean up read-only directories.

This isn't perfect: for example, a `WT_ASSERT` firing during a Python test will take down the interpreter, so the cleanup step won't run.  So I still think we should change the Jenkins jobs to cope with this case (by adding `chmod -R u+w` before `git cleanup`).  My goal here is to avoid most annoyances when running the tests manually.

File: dist/api_data.py [hunk 1]
```diff
@@ -478,6 +478,10 @@ def __cmp__(self, other):
                 vary depending on the current eviction load''',
                 min=1, max=20),
             ]),
+    Config('readonly', 'false', r'''
+        open connection in read-only mode.  The database must exist.  All
+        methods that may modify a database are disabled.''',
```
Comment (agorrod) [inline dist/api_data.py]: It feels to me as though read only deserves a page in the reference guide, with an @ref from here. I'm not inclined to document all the API restrictions in detail, but a general overview would be good. Some examples of things I think would be worth mentioning:
* What happens if logging was turned on
* What happens if recovery was required
* What happens if there are LSM trees present

File: src/lsm/lsm_tree.c [hunk 2]
```diff
@@ -420,26 +420,32 @@ __wt_lsm_tree_create(WT_SESSION_IMPL *session,
...
+		    session, buf->data, buf->size, &lsm_tree->file_config));
+
+		/* Create the first chunk and flush the metadata. */
+		WT_ERR(__wt_lsm_meta_write(session, lsm_tree));
+
+		/* Discard our partially populated handle. */
+		ret = __lsm_tree_discard(session, lsm_tree, false);
+		lsm_tree = NULL;
+	} else {
+		F_CLR(lsm_tree, WT_LSM_TREE_MERGES);
```
Comment (agorrod) [inline src/lsm/lsm_tree.c]: Would it be reasonable to add an assertion to `__wt_lsm_manager_push_entry` that it isn't called in read only mode?

File: src/conn/conn_api.c [hunk 3]
```diff
@@ -1448,8 +1473,26 @@ __conn_single(WT_SESSION_IMPL *session, const char *cfg[])
...
-	WT_ERR(__wt_open(session,
-	    WT_SINGLETHREAD, is_create || exist, false, 0, &conn->lock_fh));
+	ret = __wt_open(session,
+	    WT_SINGLETHREAD, is_create || exist, false, 0, &conn->lock_fh);
+
+	/*
+	 * If this is a read-only connection and we cannot grab the lock
+	 * file, check if it is because there is not write permission or
+	 * if the file does not exist.  If so, then ignore the error.
+	 * XXX Ignoring the error does allow multiple read-only
```
Comment (agorrod) [inline src/conn/conn_api.c]: This is something else I'd like described in a documentation section.

File: src/conn/conn_api.c [hunk 4]
```diff
@@ -1448,8 +1473,26 @@ __conn_single(WT_SESSION_IMPL *session, const char *cfg[])
...
+	 * connections to exist at the same time on a read-only directory.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY) &&
+	    (ret == EACCES || ret == ENOENT)) {
```
Comment (agorrod) [inline src/conn/conn_api.c]: Are these error codes consistent across different platforms?

File: src/conn/conn_api.c [hunk 5]
```diff
@@ -1448,8 +1473,26 @@ __conn_single(WT_SESSION_IMPL *session, const char *cfg[])
...
+		/*
+		 * If we got an expected permission or non-existence error
+		 * then skip the byte lock.
+		 */
+		ret = 0;
+		goto open_wt;
```
Comment (agorrod) [inline src/conn/conn_api.c]: Could you structure this as an if/else instead of a goto?

File: dist/api_data.py (hunk 1 above)
Comment (sueloverso) [inline dist/api_data.py]: Yes.  I do still need to do that.  Michael and I had discussed that it needs its own page in the Ref Guide.  Good suggestion to add the ref there too.

File: src/conn/conn_api.c [hunk 6]
```diff
@@ -1898,8 +1977,8 @@ wiredtiger_open(const char *home, WT_EVENT_HANDLER *event_handler,
 	    (int)sizeof(version), ENOMEM);
 	__conn_config_append(cfg, version);
 
-	/* Ignore the base_config file if we config_base set to false. */
-	if (config_base_set)
+	/* Ignore the base_config file if config_base_set is false. */
+	if (config_base_set || F_ISSET(conn, WT_CONN_READONLY))
```
Comment (agorrod) [inline src/conn/conn_api.c]: I thought this would fall out from the configuration string override. Or does that only stop us attempting to create a base configuration file?

Comment (agorrod) [inline src/conn/conn_api.c]: Oh, drat. I should have read further. I see that there is a chicken/egg thing going on here.

File: src/conn/conn_log.c [hunk 7]
```diff
@@ -133,8 +133,13 @@ __logmgr_config(
 		FLD_SET(conn->log_flags, WT_CONN_LOG_RECOVER_ERR);
 
 	WT_RET(__wt_config_gets(session, cfg, "log.zero_fill", &cval));
-	if (cval.val != 0)
+	if (cval.val != 0) {
+		if (F_ISSET(conn, WT_CONN_READONLY))
```
Comment (agorrod) [inline src/conn/conn_log.c]: Why isn't this handled in the configuration override?

File: src/session/session_api.c [hunk 8]
```diff
@@ -13,6 +13,40 @@ static int __session_snapshot(WT_SESSION *, const char *);
 static int __session_rollback_transaction(WT_SESSION *, const char *);
 
 /*
+ * __wt_session_notsup_cfg --
+ *	Unsupported session actions that have a signature of a config string.
+ */
+int
+__wt_session_notsup_cfg(
```
Comment (agorrod) [inline src/session/session_api.c]: It's a nit, but we generally use `config` in function names rather than `cfg`

**Comment by agorrod** [comment]
@sueloverso I've gone through this, it's looking good. I've noted some requests for documentation, nits and specific questions about why error checks are where they are.

File: src/conn/conn_log.c (hunk 7 above)
Comment (sueloverso) [inline src/conn/conn_log.c]: Some configs get overwritten (in the override function) and some, if specified, are conflicting and get an error.  In particular, those that ask for specific writing operations like zero-fill does, is in conflict and gives an error rather than silently changing the value.  Also particularly those where the conflict is caused by explicit request since it is not the default.

File: src/lsm/lsm_tree.c (hunk 2 above)
Comment (sueloverso) [inline src/lsm/lsm_tree.c]: done

File: src/conn/conn_api.c (hunk 4 above)
Comment (sueloverso) [inline src/conn/conn_api.c]: Good catch.  I'm going to need to make WT codes.

File: src/conn/conn_api.c (hunk 5 above)
Comment (sueloverso) [inline src/conn/conn_api.c]: done

File: src/session/session_api.c [hunk 9]
```diff
@@ -13,6 +13,39 @@ static int __session_snapshot(WT_SESSION *, const char *);
...
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)wt_session;
+	WT_UNUSED(uri);
+	WT_UNUSED(config);
+
+	WT_RET_MSG(session, ENOTSUP, "Unsupported session method");
+}
+
+/*
```
Comment (keithbostic) [inline src/session/session_api.c]: @sueloverso, you don't need two functions, you can do it with a cast (see `cur_std.c` for an example). It doesn't much matter (there are only two after all), FYI if there are ever more.

File: src/session/session_api.c [hunk 10]
```diff
@@ -546,6 +579,9 @@ __session_log_printf(WT_SESSION *wt_session, const char *fmt, ...)
 	session = (WT_SESSION_IMPL *)wt_session;
 	SESSION_API_CALL_NOCONF(session, log_printf);
 
+	if (F_ISSET(S2C(session), WT_CONN_READONLY))
+		WT_ERR_MSG(session, ENOTSUP, "Unsupported session method");
+
```
Comment (keithbostic) [inline src/session/session_api.c]: @sueloverso, why not:
```
WT_ERR(session, __wt_session_notsup(session));
```

File: src/session/session_api.c (hunk 9 above)
Comment (sueloverso) [inline src/session/session_api.c]: Thanks.  That cleaned up a bunch of code.

File: src/session/session_api.c (hunk 10 above)
Comment (sueloverso) [inline src/session/session_api.c]: This is no longer relevant with the fix above.

**Comment by sueloverso** [comment]
@michaelcahill and @agorrod this branch is now fully complete and ready to merge.  @keithbostic also did a review pass.  I added wtperf support and two workloads (based off the eviction ones because I wanted to make sure eviction and readonly played together).  That change has not been reviewed by anyone yet and should be.  Once the branch is merged I will add it into Jenkins to run/plot with the other eviction jobs to make sure it is regularly run.

File: src/include/wiredtiger.in [hunk 11]
```diff
@@ -1853,6 +1853,10 @@ struct __wt_connection {
 	 * thread uses a session handle from the configured session_max., an
 	 * integer between 3 and 20; default \c 4.}
 	 * @config{ ),,}
+	 * @config{readonly, open connection in read-only mode.  The database
+	 * must exist.  All methods that may modify a database are disabled.
+	 * See @ref readonly for more information., a boolean flag; default \c
+	 * false.}
```
Comment (agorrod) [inline src/include/wiredtiger.in]: Is it intentional to allow reconfigure to accept the readonly flag?

File: bench/wtperf/runners/evict-btree-readonly.wtperf [hunk 12]
```diff
@@ -0,0 +1,12 @@
+# wtperf options file: evict btree configuration
```
Comment (agorrod) [inline bench/wtperf/runners/evict-btree-readonly.wtperf]: Is this a performance test, or more of a functional test? If it's a performance test is it going to test something different to the non-read-only evict btree (now or in the future)?

File: src/conn/conn_api.c [hunk 13]
```diff
@@ -1109,6 +1109,28 @@ __conn_config_append(const char *cfg[], const char *config)
...
+ *	Append an entry to a config stack that overrides some settings
+ *	when read-only is configured.
+ */
+static void
+__conn_config_readonly(const char *cfg[])
+{
+	const char *readonly;
+
+	/*
+	 * Override certain settings.  Other settings at odds will return
```
Comment (agorrod) [inline src/conn/conn_api.c]: Can we better define "certain" here? Your comment below makes it sound as though the intent is for this string to cover cases where the default settings would conflict - is that a reasonable general statement?

File: src/os_win/os_errno.c [hunk 14]
```diff
@@ -32,11 +33,28 @@ __wt_map_error_to_windows_error(int error) {
...
  */
 static int
-__wt_map_windows_error_to_error(DWORD winerr) {
+__wt_map_windows_error_to_error(DWORD winerr)
+{
 	return (winerr + windows_error_offset);
 }
 
 /*
+ * __wt_map_error_rdonly --
```
Comment (agorrod) [inline src/os_win/os_errno.c]: @sueloverso I think it'd be nicer to have a generic error mapping function in `os_windows/os_errno.c` than something specific to read only mode. Maybe a translation map similar to: https://hackage.haskell.org/package/base-4.3.0.0/src/cbits/Win32Utils.c

Then I'd expect all APIs in os_windows to translate their error returns into posix versions prior to returning. I'd expect this to replace the current `map_windows_error` methods.

File: src/session/session_api.c [hunk 15]
```diff
@@ -1423,6 +1435,37 @@ __open_session(WT_CONNECTION_IMPL *conn,
...
+	/*
+	 * Disable some methods if this is a read-only connection.  Group
+	 * them by call signature.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY)) {
+		wt_session = &session_ret->iface;
+		wt_session->checkpoint =
+		    (int (*)(WT_SESSION *, const char *))__wt_session_notsup;
+		wt_session->compact =
+		    (int (*)(WT_SESSION *, const char *, const char *))__wt_session_notsup;
```
Comment (agorrod) [inline src/session/session_api.c]: @keithbostic Can we rely on the compiler knowing how much stack space is used? Casting function pointers seems worrying to me - I've been bitten by passing incorrectly sized variables through to printf more often than I care to admit..

Comment (keithbostic) [inline src/session/session_api.c]: @agorrod, I don't see why not as long as the compiler pops based on the caller's arguments, which I think it has to do. I might be missing what you're saying: can you sketch out an example of how this might fail? (We also do this in the cursor code, see `cur_std.c`.)

File: src/os_win/os_errno.c (hunk 14 above)
Comment (sueloverso) [inline src/os_win/os_errno.c]: I think that is a larger project with different requirements, potentially.  Readonly needs to detect two very specific situations.  I'd rather not delay this work for a general windows error discussion.  Can we create a new ticket for that?  It isn't clear to me that translating all errors to posix is what a Windows programmer would expect.

File: src/include/wiredtiger.in (hunk 11 above)
Comment (sueloverso) [inline src/include/wiredtiger.in]: Good catch.  I have moved it.

File: bench/wtperf/runners/evict-btree-readonly.wtperf (hunk 12 above)
Comment (sueloverso) [inline bench/wtperf/runners/evict-btree-readonly.wtperf]: I would say it is both.  I specifically chose eviction to make sure there are no surprising interactions when the cache fills up.  But I also want to make sure that readonly mode does not impact performance (potentially it could be a slight improvement if we run fewer threads or internal threads do less work, I guess).

File: src/session/session_api.c (hunk 15 above)
Comment (agorrod) [inline src/session/session_api.c]: > as long as the compiler pops based on the caller's arguments, which I think it has to do

That's exactly my concern. Stackoverflow seems to answer my question:
http://stackoverflow.com/questions/559581/casting-a-function-pointer-to-another-type

tl;dr The behavior is undefined in the C standard, but likely to work most places. I don't feel strongly about it, it's definitely cleaner to have less duplication of the not supported functions.

Comment (sueloverso) [inline src/session/session_api.c]: Hmm, that article also says this - which sounds exactly like what we're doing unless I'm reading it wrong:

Calling a function pointer boils down to pushing the arguments on the stack and doing an indirect jump to the function pointer target, and there's obviously no notion of types at the machine code level.

Things you definitely can't do:

    Cast between function pointers of different calling conventions. You will mess up the stack and at best, crash, at worst, succeed silently with a huge gaping security hole.

Comment (keithbostic) [inline src/session/session_api.c]: @sueloverso, I don't think we're using different calling conventions (unless the compiler were to choose from multiple calling conventions, perhaps based on the function signature).  I suppose it's theoretically possible, but I've never seen a compiler that used multiple calling conventions in the same compile/build, and finding one would surprise me a lot.

And it would seem to be a roadblock to linking with 3rd party libraries?

Comment (sueloverso) [inline src/session/session_api.c]: I don't have enough real knowledge in this area to even be dangerous.  I certainly prefer it, stylistically.  You've convinced me.  Alex can merge once he's good.

**Comment by agorrod** [comment]
Thanks @sueloverso lgtm, I've just updated Jenkins so that our jobs force the permissions on build_posix before running. I'll let that filter through to make sure I got it right before merging.
